include/vapaee/pool/utils.hpp

```diff
@@ -0,0 +1,149 @@
+                    to_reserve_ex, from_reserve_ex + quantity_ex);
+
+                asset conversion_ex = asset_multiply(quantity, rate);
+                asset conversion = asset_change_precision(
+                    conversion_ex, to_reserve.symbol.precision());
+
+                return make_tuple(conversion, rate);
+            }
+
+            bool pool_exists(symbol_code A, symbol_code B) {
+                pools pool_markets(contract, contract.value);
+                auto sym_index = pool_markets.get_index<"symbols"_n>();
+                return sym_index.find(symbols_get_index(A, B)) != sym_index.end();
+            }
+
+            void create_fund_attempt(name funder, uint64_t market_id) {
+                require_auth(funder);
+                pools pool_markets(contract, contract.value);
+                auto pool_it = pool_markets.find(market_id);
+                check(pool_it != pool_markets.end(), ERR_POOL_NOT_FOUND);
+                check(is_account(funder), ERR_ACCOUNT_NOT_FOUND);
+
+                fund_attempts funding_attempts(contract, contract.value);
+                funding_attempts.emplace(contract, [&](auto & row) {
+                    row.auto_id = funding_attempts.available_primary_key();
+                    row.id = pack(funder.value, market_id);
+                    row.commodity = asset(0, pool_it->commodity_reserve.symbol);
+                    row.currency = asset(0, pool_it->currency_reserve.symbol);
+                });
+            }
+
+            void maybe_end_fund_attempt(name funder, uint64_t market_id) {
+                fund_attempts funding_attempts(contract, contract.value);
+                auto pack_index = funding_attempts.get_index<"idpacked"_n>();
+                auto fund_it = pack_index.find(pack(funder.value, market_id));
+                
+                check(fund_it != pack_index.end(), ERR_ATTEMPT_NOT_FOUND);
+
+                asset commodity_ex = asset_change_precision(
+                        fund_it->commodity, ARITHMETIC_PRECISION);
+                asset currency_ex = asset_change_precision(
+                        fund_it->currency, ARITHMETIC_PRECISION);
+
+                asset fund_rate = asset_divide(commodity_ex, currency_ex);
+                
+                if (fund_rate == get_pool_rate(fund_it->get_pool_id())) {
+                    // TODO: emit fee participation token
+                    
+                    pools pool_markets(contract, contract.value);
+                    auto pool_it = pool_markets.find(fund_it->get_pool_id());
+                    check(pool_it != pool_markets.end(), ERR_POOL_NOT_FOUND);
+
+                    pool_markets.modify(pool_it, contract, [&](auto &row) {
+                        row.commodity_reserve += fund_it->commodity;
+                        row.currency_reserve += fund_it->currency;
+                    });
+
+                    pack_index.erase(fund_it);
+                }
+            }
```

Comment:
Turn this "maybe" into an "always" (not talking about the title but the logic inside this function).

Pseudocode for this function:
```c++
if (exist_pool_for_market(market_id)) {

    // take the current ratio for the pool
    pool_rate = get_pool_rate(fund_it->get_pool_id());
    fund_rate = get_fund_rate(funt_it);
    current_liquidity = get_current_liquidity_total_credits(market_id);

    if (pool_rate > fund_rate) {
        // There's a surplus of currency in fund attempt
        diff = calculate_surplus_of_currency(pool_rate, fund_it);
        fund_pool(market_id, fund_it->commodity);
        fund_pool(market_id, fund_it->currency - diff);
        // return surplus tokens to funder
        transfer_tokens(contract, funder, diff, "Surplus in funding attempt to market " + market.toString());

    } else if (pool_rate < fund_rate) {
        // There's a surplus of commodity in fund attempt
        diff = calculate_surplus_of_commodity(pool_rate, fund_it);
        fund_pool(market_id, fund_it->commodity - diff);
        fund_pool(market_id, fund_it->currency);
        // return surplus tokens to funder
        transfer_tokens(contract, funder, diff, "Surplus in funding attempt to market " + market.toString());

    } else {
        // use the total amount of tokens provided for the fund attempt
        fund_pool(market_id, fund_it->commodity);
        fund_pool(market_id, fund_it->currency);
    }

    // gives the investor a number of credits
    // equivalent to the percentage of the liquidity provided
    // relative to the existing pool liquidity    
    liquidity_provided = get_current_liquidity_total_credits(market_id) - current_liquidity;
    creadits_for_investor(funder, market_id, liquidity_provided);

} else {
    // pool does not exist. So we created now
    create_pool_for_market(market_id);

    // use the total amount of tokens provided for the fund attempt
    fund_pool(market_id, fund_it->commodity);
    fund_pool(market_id, fund_it->currency);

    // gives the investor a number of credits
    // equivalent to the 100% of the liquidity of this pool
    total_liquidity = get_current_liquidity_total_credits(market_id);
    creadits_for_investor(funder, market_id, total_liquidity);
}

```